pkg/cloud/services/compute/instance.go

```diff
@@ -345,7 +346,19 @@ func (s *Service) createInstanceImpl(eventObject runtime.Object, openStackCluste
 
 	serverCreateOpts = applyServerGroupID(serverCreateOpts, instanceSpec.ServerGroupID)
 
-	server, err = s.getComputeClient().CreateServer(keypairs.CreateOptsExt{
+	// Filter supported versions based on features used
+	recognizedMicroversions := clients.NovaSupportedVersions
+	supportedMicroversions := []*utils.Version{}
+	if len(instanceSpec.Tags) > 0 {
+		for i := range recognizedMicroversions {
+			if recognizedMicroversions[i].ID == "2.1" {
+				continue
+			}
```

Comment:
Good question! I'm not sure if this is how we should do it, but I can explain what I'm trying to do here.
Before we [set the minimum version](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/commit/af5069cd715e401a5cbc261500286d4217267b1a) we didn't set any version at all. If I understand correctly that would mean the same thing as setting 2.1.

> If no version is specified then the API will behave as if a version request of v2.1 was requested.

From https://docs.openstack.org/nova/latest/reference/api-microversion-history.html

So what I'm trying to do here is to allow fallback to the default if possible. If tags are used, then we require 2.53 minimum so then I remove 2.1 from the versions that it can pick from. Probably a better way to do it would be to convert the versions to floats and then remove anything that is < 2.53, but for now I'm just playing with it.

Reply:
Incidentally however we do it, I'd probably go with ints rather than floats and drop the `2.`.

I wonder if this could be something like:
```golang
compute := s.getComputeClient()
if len(instanceSpec.Tags) > 0 {
  err := compute.RequireMicroVersion(NovaTagging)
  if err != nil {
    ...tagging not supported...
  }
}
```

We could have similar for neutron extensions.

Thoughts:
* how can we inform the user up front, if their configuration requires more than their cloud supports?
* If we can't do it up front, how do we make the best UX we can in the failure case?
* How do we do the 'we need the multi-attach microversion' thing, because we can't detect that from k8s objects?

On the latter point: we could pre-fetch the specified volume-type, or infer the default volume type (don't know if this is possible) and check if multi-attach is set. Or we could react to a failure by trying again with a higher microversion, but the granularity of OpenStack error code likely makes this messy.

Reply:
> how can we inform the user up front, if their configuration requires more than their cloud supports?

I think upfront is impossible here. We don't do any API calls in the webhook and for good reason, so I don't see how we would be able to do it unfortunately.

> If we can't do it up front, how do we make the best UX we can in the failure case?

Probably the best we can do here is to bubble up the error to the CAPI level where it will be visible on the Machine or the Cluster.

>   How do we do the 'we need the multi-attach microversion' thing, because we can't detect that from k8s objects?
> On the latter point: we could pre-fetch the specified volume-type, or infer the default volume type (don't know if this is possible) and check if multi-attach is set. Or we could react to a failure by trying again with a higher microversion, but the granularity of OpenStack error code likely makes this messy.

Now we are getting to the core of the issue! I *think* the reasonable thing to do is to keep a list of supported versions with higher priority for higher versions (as seen in this PR). That way the highest supported version would be picked (I hope), which should support as many features as possible. My thinking is that this would solve the issue without complicating the code with extra checks, errors and retries.

However, it builds on the assumption that picking the highest supported version is always best and I'm not completely sure if this is true. Could there be a case similar to the multi-attach volume situation, where OpenStack is configured in a way that a microversion is in theory supported, but in practice a *lower* version is needed because of `specific-volume-type-or-similar` set up by the admins? I want to think the answer is no and that it would be safe to always pick the highest available versions.